servant-client/src/Servant/Client/Generic.hs

```diff
@@ -0,0 +1,82 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DefaultSignatures #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+module Servant.Client.Generic
+  ( ClientLike(..)
+  , genericMkClient
+  ) where
+
+import Generics.SOP   (Generic, I(..), NP(..), NS(Z), Rep, SOP(..), to)
+import Servant.API    ((:<|>)(..))
+import Servant.Client (ClientM)
+
+-- | This class allows us to match client structure with client functions
+-- produced with 'client' without explicit pattern-matching.
+--
+-- The client structure needs a 'Generics.SOP.Generic' instance.
+--
+-- Example:
+--
+-- > type API
+-- >     = "foo" :> Capture "x" Int :> Get '[JSON] Int
+-- >  :<|> "bar" :> QueryParam "a" Char :> QueryParam "b" String :> Post '[JSON] [Int]
+-- >  :<|> Capture "nested" Int :> NestedAPI
+-- >
+-- > type NestedAPI
+-- >     = Get '[JSON] String
+-- >  :<|> "baz" :> QueryParam "c" Char :> Post '[JSON] ()
+-- >
+-- > data APIClient = APIClient
+-- >   { getFoo         :: Int -> ClientM Int
+-- >   , postBar        :: Maybe Char -> Maybe String -> ClientM [Int]
+-- >   , mkNestedClient :: Int -> NestedClient
+-- >   } deriving GHC.Generic
+-- >
+-- > instance Generic.SOP.Generic APIClient
+-- > instance (Client API ~ client) => ClientLike client APIClient
+-- >
+-- > data NestedClient = NestedClient
+-- >  { getString ::  ClientM String
+-- >  , postBaz   :: Maybe Char -> ClientM ()
+-- >  } deriving GHC.Generic
+-- >
+-- > instance Generic.SOP.Generic
+-- > instance (Client NestedAPI ~ client) => ClientLike client NestedClient
+-- >
+-- > mkAPIClient :: APIClient
+-- > mkAPIClient = mkClient (client (Proxy :: Proxy API))
+class ClientLike client custom where
```

Comment:
> I think I'd intuitively prefer accepting the non-nested definition for left-nested occurrences of `:<|>` instead of requiring the nested one. This would be easy to do, and seems a bit more symmetric to me.

That's actually not that easy to do. Now we take only the head of the type level list, but if we go left and right, we'd have to use `xs ++ ys` which can't be used in an instance head. So we'd need to define a separate class with associated type family to split `zs` into `xs` and `ys` based on `a :<|> b` client. This is doable, of course, just messy.

This approach also makes `a :<|> b` decide how the custom client should be split... which is probably not a good idea, although I am not 100% sure. One concern is that `a :<|> b :<|> c` client can be represented by either 3 fields in a client record, or 1 field (client sub-record). I suppose that means that if we flatten the nested `:<|>` on the left, sometimes GHC wouldn't be able to resolve the types. To eliminate this issue we could add `instance TypeError ... => ClientLike (a :<|> b) '[x]` so that `:<|>` would have to be flattened always...

So I guess, we can do it, but I am not sure yet that we should.

> I don't understand the argument about N-tuples here. I think @jkarni is right. Flattening is an orthogonal operation. If you have a type-level computation Flatten that turns your API into flattened form, then you can do the entire construction on Flatten API, and then successfully map to a flat datatype.

Here's a simple example of a client and flattened client. Which instances should I add to `ClientLike` to allow `mkClient :: Client -> FlattenedClient`? I doesn't work out of the box right now and I don't see an immediate solution here (accept the one that is essentially its own type class for flattening, [such as one here](http://hackage.haskell.org/package/smsaero-0.4/docs/SMSAero-Utils.html))

```haskell
type Client = Maybe Int -> Maybe String -> (ClientM Int :<|> ClientM String)

type FlattenedClient
    = (Maybe Int -> Maybe String -> ClientM Int)
 :<|> (Maybe Int -> Maybe String -> ClientM String)
```